### PR TITLE
Update CooldownButton.cs to my version

### PR DESCRIPTION
**Pull request by Pandapip1**
Reference: https://gist.github.com/Pandapip1/b306f77187ce2f25e73be2560da4aa5f

### CODE REVIEW

**Comment by Pandapip1** [comment]
Improvements:
- Takes byte[] as image (instead of string)
- Uses delegates to determine if a user can use the button or not
- Cooldown does not decrease while on cams / in vent / etc...
- Cooldown resets on meeting

File: Essentials/UI/CooldownButton.cs [hunk 1]
```diff
@@ -1,295 +1,226 @@
...
+        public float Timer = 0f;
+        public float EffectDuration = 0f;
+        public bool isEffectActive;
+        public bool hasEffectDuration;
+        public bool enabled = true;
+        public UseTester useTester;
+        private byte[] image;
+        private Action OnClick;
+        private Action OnEffectEnd;
+        private HudManager hudManager;
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:72]: This is unnecessary.

File: Essentials/UI/CooldownButton.cs [hunk 2]
```diff
@@ -1,295 +1,226 @@
...
-        private bool IsDisposed;
-
-        /// <summary>
-        /// Call in HudManager.Start
-        /// </summary>
-        private CooldownButton(Assembly asm, Action onClick, float cooldown, string embeddedResourcePath/*, float pixelsPerUnit*/, Vector2 positionOffset, ButtonCategory category, bool hasEffectDuration, float effectDuration, Action onEffectEnd)
+        public float MaxTimer = 0f;
+        public float Timer = 0f;
+        public float EffectDuration = 0f;
+        public bool isEffectActive;
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:65]: Class, method, property and non-local field names should always be PascalCase.

File: Essentials/UI/CooldownButton.cs [hunk 3]
```diff
@@ -1,295 +1,226 @@
...
+        public bool hasEffectDuration;
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:66]: Class, method, property and non-local field names should always be PascalCase.

File: Essentials/UI/CooldownButton.cs [hunk 4]
```diff
@@ -1,295 +1,226 @@
...
+        public bool enabled = true;
+        public UseTester useTester;
+        private byte[] image;
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:69]: Class, method, property and non-local field names should always be PascalCase.

File: Essentials/UI/CooldownButton.cs [hunk 5]
```diff
@@ -1,295 +1,226 @@
...
+        private Action OnClick;
+        private Action OnEffectEnd;
+        private HudManager hudManager;
+        private bool canUse;
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:73]: Class, method, property and non-local field names should always be PascalCase.

File: Essentials/UI/CooldownButton.cs [hunk 6]
```diff
@@ -1,295 +1,226 @@
...
+
+        public CooldownButton(Action OnClick, float Cooldown, byte[] image, Vector2 PositionOffset, UseTester useTester, HudManager hudManager, float EffectDuration, Action OnEffectEnd)
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:75]: Local fields should always be camelCase.

File: Essentials/UI/CooldownButton.cs [hunk 7]
```diff
@@ -1,295 +1,226 @@
...
-            Stream resourceStream = asm.GetManifestResourceStream(embeddedResourceFullPath);
-
-            KillButtonManager = Object.Instantiate(HudManager.Instance.KillButton, HudManager.Instance.transform);
-
-            //StartColorButton = killButtonManager.renderer.color;
-            StartColorText = KillButtonManager.TimerText.Color;
-
-            KillButtonManager.gameObject.SetActive(true);
-            KillButtonManager.renderer.enabled = true;
+        public CooldownButton(Action OnClick, float Cooldown, byte[] image, Vector2 PositionOffset, UseTester useTester, HudManager hudManager)
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:92]: You can use `: this(...)` to call a different constructor, there's no need to the constructor code.

File: Essentials/UI/CooldownButton.cs [hunk 8]
```diff
@@ -1,295 +1,226 @@
...
+            this.useTester = useTester;
+            MaxTimer = Cooldown;
+            Timer = MaxTimer;
+            this.image = image;
+            hasEffectDuration = false;
+            buttons.Add(this);
+            Start();
+        }
 
+        private void Start()
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:106]: Did something similar in my refactor, this is sadly missing null checks and a reason for existing.

File: Essentials/UI/CooldownButton.cs [hunk 9]
```diff
@@ -1,295 +1,226 @@
...
+            }
         }
-
-        /// <summary>
-        /// Call in HudManager.Start
-        /// </summary>
-        public CooldownButton(Action onClick, float cooldown, string imageEmbededResourcePath, Vector2 positionOffset, ButtonCategory category) : this(Assembly.GetCallingAssembly(), onClick, cooldown, imageEmbededResourcePath, positionOffset, category, false, 0F, null)
+        public bool CanUse()
         {
+            if (PlayerControl.LocalPlayer.Data == null) return false;
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:136]: `PlayerControl.LocalPlayer` can also be null, resulting an exception, to check for both, use `PlayerControl.LocalPlayer?.Data`.

File: Essentials/UI/CooldownButton.cs [hunk 10]
```diff
@@ -1,295 +1,226 @@
...
+            return true;
         }
-
-        public bool IsUsable()
+        public static void HudUpdate()
         {
-            if (PlayerControl.LocalPlayer?.Data == null) return false;
-
-            switch (Category)
+            buttons.RemoveAll(item => item.killButtonManager == null);
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:142]: Nullifies the reason for `Start()`'s existence since the `CooldownButton` wrapper object becomes irrelevant once the wrapped `KillButtonManager` instance no longer exists. Yes this is an issue with the existing code too, but it is changing.

File: Essentials/UI/CooldownButton.cs [hunk 11]
```diff
@@ -1,295 +1,226 @@
...
-
-        [HarmonyPatch(typeof(HudManager), nameof(HudManager.Update))]
-        private static class HudUpdatePatch
-        {
-            public static void Postfix()
-            {
-                Buttons.RemoveAll(item => item.KillButtonManager == null);
-
-                for (int i = 0; i < Buttons.Count; i++)
+                catch (NullReferenceException)
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:150]: Happens due to the improper `PlayerControl.LocalPlayer.Data` check mentioned above. Corrected in my refactor.

File: Essentials/UI/CooldownButton.cs [hunk 12]
```diff
@@ -1,295 +1,226 @@
...
                     OnEffectEnd();
                 }
             }
             else
             {
-                if (CanUse) Timer -= Time.deltaTime;
-
-                KillButtonManager.renderer.color = new Color(1f, 1f, 1f, 0.3f);
+                if (canUse)
+                    if (PlayerControl.LocalPlayer.CanMove)
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:174]: Why nest `if`s?

File: Essentials/UI/CooldownButton.cs [hunk 13]
```diff
@@ -1,295 +1,226 @@
...
-                    catch (NullReferenceException)
-                    {
-                        System.Console.WriteLine("[WARNING] NullReferenceException from HudUpdate().CanUse(), if theres only one warning its fine");
-                    }
+                    System.Console.WriteLine("[WARNING] NullReferenceException from HudUpdate().CanUse(), if theres only one warning its fine");
                 }
             }
         }
-
         private void Update()
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:156]: Null checks?

File: Essentials/UI/CooldownButton.cs [hunk 14]
```diff
@@ -1,295 +1,226 @@
...
-                KillButtonManager.renderer.material.SetFloat("_Desat", 0f);
-
-                KillButtonManager.SetCoolDown(Timer, MaxTimer);
+                killButtonManager.renderer.material.SetFloat("_Desat", 0f);
+                killButtonManager.SetCoolDown(Timer, MaxTimer);
             }
         }
-
-        public void ApplyCooldown()
+        internal delegate bool d_LoadImage(IntPtr tex, IntPtr data, bool markNonReadable);
```
Comment (DorCoMaNdO) [inline Essentials/UI/CooldownButton.cs:186]: Why is this here? The implementation in Reactor is being called anyway?

**Review by DorCoMaNdO — COMMENTED** [review]
I've got a proper refactor underway that would make buttons simpler to create and use, as well as make more consistent in behavior, but I do want to point out issues with your version.

File: Essentials/UI/CooldownButton.cs (hunk 12 above)
Comment (DorCoMaNdO) [reply Essentials/UI/CooldownButton.cs:174]: Also this prevents effect durations from ticking down.

**Comment by Pandapip1** [comment]
Thank you for taking the time to review the code ;)
I will apply your suggestions.

**Comment by DorCoMaNdO** [comment]
I've pushed the first pass of my refactor so I'm closing this, if you have any feedback for my implementation please do open an issue or PR.
